### ISSUE TITLE
spdk_sock_connect blocks  reactor running

### ISSUE THREAD

**Issue by winglq** [spdk/spdk#2062]
spdk_posix_sock_create(posix.c) sets socket to O_NONBLOCK until connect is successful. If the connection target is down(power off), connect will bock until the operation is timeout(75s in linux?). In this situation the reactor will be blocked, if spdk_sock_connect is called by reactor_run. 
Will this behavior be changed to a non block call, or any suggestion on how should I call spdk_sock_connect?
Thanks.

**Comment by jimharris** [comment]
Hi @winglq,

We should change this to O_NONBLOCK.  But that will require quite a few other changes - for example, making the sock library create/connect API asynchronous, and changing the nvme driver's tcp transport to handle an asynchronous connect.  There is already work in progress on the nvme driver's tcp transport which would make this easier.

Thanks,

-Jim

**Comment by winglq** [comment]
Thanks Jim.
Currently I have to create a pthread to deal with the connect.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @optimistyzy. log: https://ci.spdk.io/public_build/autotest-per-patch_55846.html

**Comment by Comphix** [comment]
Looks like we have several patches for the nvme/tcp async operations are merged.

**Comment by jimharris** [comment]
> Looks like we have several patches for the nvme/tcp async operations are merged.

Yes, and the rest of the series is here: https://review.spdk.io/gerrit/c/spdk/spdk/+/8607

Honestly, I think the parts of the series that we really need for async spdk_sock_connect have already been merged.  @winglq are you interested in moving this forward in upstream SPDK?

**Comment by winglq** [comment]
I am happy to contribute. Please guide me on how to move this forward and I hope there is no urgent time goal.
Thanks.

**Comment by Comphix** [comment]
Thanks @winglq , you could follow below link to contribute the patch.

https://spdk.io/development/

**Comment by winglq** [comment]
I have uploaded the patch. I used perf(nvme-of target tcp protocol) to test it with following cases:
1. good path
2. connect failure
3. connect timeout

Please help to review, thanks.

https://review.spdk.io/gerrit/c/spdk/spdk/+/9508

**Comment by jimharris** [comment]
Hi @winglq,

Thanks for your patch!  I added a few comments on Gerrit, and also kicked our CI to run tests against your patch.  I'll do a more thorough review on the functional parts of the patch once it's gone through CI.

-Jim

**Comment by jimharris** [comment]
Hi @winglq,

scan-build found a dead assignment.  If you aren't familiar with how to check CI results, go to your Gerrit URL, look for the -1 CI vote, and you'll see a link in there that takes you to CI results.  From there click on scanbuild link that shows failure, and then on the "scan-build" link that comes next.

-Jim

**Comment by jimharris** [comment]
Hi @winglq,

This patch slipped through the maintainers workflows (including mine).  If you're still interesting on getting this patch moved forward that's great, but if not, we will leave this open in case someone else wants to pick this up and move it to closure.

I'll be posting some additional feedback on the patch here shortly in Gerrit.

-Jim

**Comment by mtzaurus** [comment]
Folks, what's the status with this issue ? Is the fix gonna be merged any time soon ?

**Comment by jimharris** [comment]
> Folks, what's the status with this issue ? Is the fix gonna be merged any time soon ?

Hi @mtzaurus,

Tiago Castro (don't have his GitHub handle) asked about this on Slack last week and tagged you.  He said he was going to take a look at the patch.  At this point the patch just needs someone to pick it up from the original author, go through reviews and get it merged.

-Jim

**Comment by ksztyber** [comment]
[Bug scrub] The original patch posted still have some issues, but marking as an enhancement.

**Comment by jkalwas** [comment]
yet another case when blocking connection might be problematic is when initiator tries to connect with the remote target and it has firewall blocking that connection

that is likely misconfiguration of the remote target however affects initiator significantly by being blocked for few (30 or 60?) seconds

**Comment by AppleTp** [comment]
Hi@jimharris

**Comment by AppleTp** [comment]
@jimharris

**Comment by jimharris** [comment]
> @jimharris
> 
> I had the same problem recently, and I'm happy to move forward with the previous patch. But before proceeding, I have the following two questions.
> 
> **What should I do with the previous patch?** Download the code and resubmit it locally? Or just use the previous author's patch?

Please download the existing patch locally, make changes as necessary, and resubmit as a new Gerrit review. Please also add reference in commit message to the original Gerrit URL.

> **To change the socket connection to asynchronous, I think there are a couple of options**
> 
> 1. Add connection options to make the asynchronous connection method optional, so that when asynchronous connections are not used by default, they do not interfere with other uses.
> 2. Add the asynchronous connection option and keep the synchronous connection as a transition for the time being to avoid excessive impact of this modification
> 3. Modify the connection method to asynchronous, no options are provided. Modifying it in this way will completely affect the use of the socket module, and will require better testing for sockets

Definitely not #3.  We need to keep synchronous APIs for now, maybe indefinitely. There should be little maintenance burden to keep the synchronous APIs - those functions can just use the async APIs internally.

We need to consider whether all impls must support async connect. I think we should make this a requirement, but it means we need to implement in both posix and uring socket modules before merging the code.

Ideally this is just another sock option, and can use the existing `spdk_sock_connect_ext()` API.  Received oriented calls on an unconnected socket just return 0 bytes until connected. Async transmit calls can just queue up the writes internally. Synchronous transmit calls just return an error - if users want async connect they need to use async transmit as well.

We should add some kind of `spdk_sock_connect_poll()` API, where caller can poll the socket one time to progress on the connection.

**Comment by AppleTp** [comment]
> > @jimharris
> > I had the same problem recently, and I'm happy to move forward with the previous patch. But before proceeding, I have the following two questions.
> > **What should I do with the previous patch?** Download the code and resubmit it locally? Or just use the previous author's patch?
> 
> Please download the existing patch locally, make changes as necessary, and resubmit as a new Gerrit review. Please also add reference in commit message to the original Gerrit URL.
> 
> > **To change the socket connection to asynchronous, I think there are a couple of options**
> > 
> > 1. Add connection options to make the asynchronous connection method optional, so that when asynchronous connections are not used by default, they do not interfere with other uses.
> > 2. Add the asynchronous connection option and keep the synchronous connection as a transition for the time being to avoid excessive impact of this modification
> > 3. Modify the connection method to asynchronous, no options are provided. Modifying it in this way will completely affect the use of the socket module, and will require better testing for sockets
> 
> Definitely not #3. We need to keep synchronous APIs for now, maybe indefinitely. There should be little maintenance burden to keep the synchronous APIs - those functions can just use the async APIs internally.
> 
> We need to consider whether all impls must support async connect. I think we should make this a requirement, but it means we need to implement in both posix and uring socket modules before merging the code.
> 
> Ideally this is just another sock option, and can use the existing `spdk_sock_connect_ext()` API. Received oriented calls on an unconnected socket just return 0 bytes until connected. Async transmit calls can just queue up the writes internally. Synchronous transmit calls just return an error - if users want async connect they need to use async transmit as well.
> 
> We should add some kind of `spdk_sock_connect_poll()` API, where caller can poll the socket one time to progress on the connection.

Thank you very much for your reply, I'll build on the previous revisions and make changes based on your comments.

**Comment by nearora-msft** [comment]
@AppleTp , I am running into this issue when I am upgrading Kubernetes cluster while I have I/O workloads running that use spdk.
I was wondering if you already have the updated patch?

**Comment by hjwsm1989** [comment]
@jimharris hi, does the community have any plan on this?

**Comment by jimharris** [comment]
> @jimharris hi, does the community has any plan on this?

Hi @hjwsm1989,

I'm not aware of anyone working on this currently. The original patch submitted by @winglq is still there for anyone interested in continuing the work started there: https://review.spdk.io/gerrit/c/spdk/spdk/+/9508.

**Comment by changpe1** [comment]
> > @jimharris hi, does the community has any plan on this?
> 
> Hi @hjwsm1989,
> 
> I'm not aware of anyone working on this currently. The original patch submitted by @winglq is still there for anyone interested in continuing the work started there: https://review.spdk.io/gerrit/c/spdk/spdk/+/9508.

I will continue improving this patch and following nvme drivers that use BLOCK connect.

**Comment by jkalwas** [comment]
This is a step in the middle of transitioning to async connect.
https://review.spdk.io/c/spdk/spdk/+/25795

**Comment by jkalwas** [comment]
... and there is the rest of the series pending testing and review.
25907: sock/posix: make connect async (non blocking) | https://review.spdk.io/c/spdk/spdk/+/25907
